Can't connect to Cloudant through proxy

Please [read these guidelines](http://ibm.biz/cdt-issue-guide) before opening an issue.

<!-- Issues will be CLOSED IMMEDIATELY if the following template is not completed. -->

## Bug Description

### 1. Steps to reproduce and the simplest code sample possible to demonstrate the issue
I'm calling `proxyURL` on my ClientBuilder, but it doesn't seem to be getting used.  Just getting a TimeoutException.  The same program, used in an environment where a proxy isn't needed, works great.  And that proxy (http://10.177.140.102:3128) works fine for other Java libraries I'm using.  Note this proxy does NOT require credentials be passed, so I'm not calling `proxyUser` or `proxyPassword` on the ClientBuilder.  Note the real app is an Open Liberty-based JAX-RS microservice running in the Tanzu Kubernetes Grid.  But I've reproduced the problem with the following simpler code:

```
import com.cloudant.client.api.ClientBuilder;
import com.cloudant.client.api.CloudantClient;
import com.cloudant.client.api.Database;
import java.io.InputStream;
import java.net.URL;
import java.nio.charset.StandardCharsets;

public class CloudantTest {
    public static void main(String[] arg) {
        if (arg.length == 1) {
            System.out.println(getEntryFromCloudant(arg[0]));
        } else {
            System.out.println("Usage: CloudantTest <id>");
        }
    }

    public static String getEntryFromCloudant(String id) {
        String response = null;
        try {
            String url = System.getenv("CLOUDANT_URL");
            if ((url == null) || url.isEmpty()) return "CLOUDANT_URL not set";

            String username = System.getenv("CLOUDANT_ID");
            if ((username == null) || username.isEmpty()) return "CLOUDANT_ID not set";

            String password = System.getenv("CLOUDANT_PASSWORD");
            if ((password == null) || password.isEmpty()) return "CLOUDANT_PASSWORD not set";

            String db = System.getenv("CLOUDANT_DB");
            if ((db == null) || db.isEmpty()) return "CLOUDANT_DB not set";

            ClientBuilder builder = ClientBuilder.url(new URL(url)).username(username).password(password);

            String proxy = System.getenv("CLOUDANT_PROXY");
            if ((proxy != null) && !proxy.isEmpty()) {
                System.out.println("Adding proxy");
                builder = builder.proxyURL(new URL(proxy));
            }

            CloudantClient client = builder.build();

            Database database = client.database(db, false);

            InputStream stream = database.find(id);
            response = new String(stream.readAllBytes(), StandardCharsets.UTF_8);
        } catch (Throwable t) {
            t.printStackTrace();
            response = t.getMessage();
        }
        return response;
    }
}
```
And I'm using this manifest in the "executable" jar to get the class path setup right:
```
Manifest-Version: 1.0
Class-Path: cloudant-client-2.19.2.jar cloudant-http-2.19.2.jar gson-2.8.2.jar commons-io-2.4.jar
Main-Class: CloudantTest
```
### 2. What you expected to happen
The connection to Cloudant should have gone through the proxy out to the internet and worked.  Here's the output from my test program above when I run it in an environment that doesn't need a proxy (like my laptop):

```
jalcorn@Johns-MBP-8 account % java -jar cloudantTest.jar 014aa25148df4eea85b15b3e81ce52d7              
{"_id":"014aa25148df4eea85b15b3e81ce52d7","_rev":"2-99b94bc1438706a4d92ad5611e8ffda9","owner":"Looper8","loyalty":"Basic","balance":40.01,"commissions":9.99,"free":0,"sentiment":"Unknown","nextCommission":9.99}

jalcorn@Johns-MBP-8 account % 
```

### 3. What actually happened
I get the following exception when running the above when `kubectl exec`' d into the Kubernetes pod (which needs the proxy to get out to the internet):

```
com.cloudant.client.org.lightcouch.CouchDbException: Error retrieving server response at https://878ff489-3ab9-4d2d-86db-016f19864439-bluemix.cloudantnosqldb.appdomain.cloud:443/account/0004422e801c4590bab285e31b769ea4.
        at com.cloudant.client.org.lightcouch.CouchDbClient.execute(CouchDbClient.java:632)
        at com.cloudant.client.org.lightcouch.CouchDbClient.executeToInputStream(CouchDbClient.java:648)
        at com.cloudant.client.org.lightcouch.CouchDbClient.get(CouchDbClient.java:389)
        at com.cloudant.client.org.lightcouch.CouchDatabaseBase.find(CouchDatabaseBase.java:148)
        at com.cloudant.client.api.Database.find(Database.java:839)
        at CloudantTest.getEntryFromCloudant(CloudantTest.java:44)
        at CloudantTest.main(CloudantTest.java:11)
Caused by: java.io.IOException: Failed to read server response from  https://878ff489-3ab9-4d2d-86db-016f19864439-bluemix.cloudantnosqldb.appdomain.cloud:443/_session
        at com.cloudant.http.internal.interceptors.CookieInterceptorBase.wrapIOException(CookieInterceptorBase.java:263)
        at com.cloudant.http.internal.interceptors.CookieInterceptorBase.makeSessionRequest(CookieInterceptorBase.java:164)
        at com.cloudant.http.internal.interceptors.CookieInterceptorBase.requestCookie(CookieInterceptorBase.java:100)
        at com.cloudant.http.internal.interceptors.CookieInterceptorBase.interceptRequest(CookieInterceptorBase.java:176)
        at com.cloudant.http.HttpConnection.execute(HttpConnection.java:291)
        at com.cloudant.client.org.lightcouch.CouchDbClient.execute(CouchDbClient.java:552)
        ... 6 more
Caused by: java.net.ConnectException: Connection timed out (Connection timed out)
        at java.base/java.net.PlainSocketImpl.socketConnect(Native Method)
        at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:399)
        at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:242)
        at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:224)
        at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
        at java.base/java.net.Socket.connect(Socket.java:609)
        at java.base/sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:289)
        at java.base/sun.net.NetworkClient.doConnect(NetworkClient.java:177)
        at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:474)
        at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:569)
        at java.base/sun.net.www.protocol.https.HttpsClient.<init>(HttpsClient.java:265)
        at java.base/sun.net.www.protocol.https.HttpsClient.New(HttpsClient.java:372)
        at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(AbstractDelegateHttpsURLConnection.java:203)
        at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1187)
        at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1081)
        at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:189)
        at java.base/sun.net.www.protocol.http.HttpURLConnection.getOutputStream0(HttpURLConnection.java:1367)
        at java.base/sun.net.www.protocol.http.HttpURLConnection.getOutputStream(HttpURLConnection.java:1342)
        at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getOutputStream(HttpsURLConnectionImpl.java:246)
        at com.cloudant.http.HttpConnection.execute(HttpConnection.java:317)
        at com.cloudant.http.internal.interceptors.CookieInterceptorBase.makeSessionRequest(CookieInterceptorBase.java:140)
        ... 10 more
```
## Environment details

- Version(s) that are affected by this issue.
    > 2.19.2
- Java version including vendor and platform
    > Eclipse OpenJ9 VM, version 11.0.12+7 (en_US)
- Using the optional okhttp dependency.
    > Yes

here you go:
```
sh-4.4$ java -Djava.util.logging.config.file=./logging.properties -jar cloudantTest.jar 014aa25148df4eea85b15b3e81ce52d7
Aug 14, 2021 3:27:31 AM com.cloudant.client.api.ClientBuilder <init>
CONFIG: URL: https://878ff489-3ab9-4d2d-86db-016f19864439-bluemix:58819704e70bb791d81eb09847e8da2d69cb492ed4fe35a4e682e38c0785e5e9@878ff
489-3ab9-4d2d-86db-016f19864439-bluemix.cloudantnosqldb.appdomain.cloud
Adding proxy
Aug 14, 2021 3:27:31 AM com.cloudant.client.api.ClientBuilder build
CONFIG: Building client using URL: https://878ff489-3ab9-4d2d-86db-016f19864439-bluemix.cloudantnosqldb.appdomain.cloud:443
Aug 14, 2021 3:27:32 AM com.cloudant.client.api.ClientBuilder build
CONFIG: Added cookie interceptor
Aug 14, 2021 3:27:32 AM com.cloudant.client.api.ClientBuilder build
CONFIG: Connect timeout: 5 MINUTES
Aug 14, 2021 3:27:32 AM com.cloudant.client.api.ClientBuilder build
CONFIG: Read timeout: 5 MINUTES
Aug 14, 2021 3:27:32 AM com.cloudant.client.api.ClientBuilder build
CONFIG: Using default GSON builder
Aug 14, 2021 3:27:32 AM com.cloudant.http.internal.ok.OkHelper <clinit>
FINE: Failed to load okhttp: okhttp3.OkUrlFactory
Aug 14, 2021 3:27:32 AM com.cloudant.client.org.lightcouch.CouchDbClient <init>
CONFIG: Using built-in HttpUrlConnection
Aug 14, 2021 3:27:32 AM com.cloudant.http.internal.DefaultHttpUrlConnectionFactory setProxy
CONFIG: Configured HTTP proxy url http://10.177.140.102:3128
Aug 14, 2021 3:27:32 AM com.cloudant.http.HttpConnection execute
FINE: c081e450-9 POST https://878ff489-3ab9-4d2d-86db-016f19864439-bluemix.cloudantnosqldb.appdomain.cloud:443/_session request
com.cloudant.client.org.lightcouch.CouchDbException: Error retrieving server response at https://878ff489-3ab9-4d2d-86db-016f19864439-bluemix.cloudantnosqldb.appdomain.cloud:443/account/014aa25148df4eea85b15b3e81ce52d7.
        at com.cloudant.client.org.lightcouch.CouchDbClient.execute(CouchDbClient.java:632)
        at com.cloudant.client.org.lightcouch.CouchDbClient.executeToInputStream(CouchDbClient.java:648)
        at com.cloudant.client.org.lightcouch.CouchDbClient.get(CouchDbClient.java:389)
        at com.cloudant.client.org.lightcouch.CouchDatabaseBase.find(CouchDatabaseBase.java:148)
        at com.cloudant.client.api.Database.find(Database.java:839)
        at CloudantTest.getEntryFromCloudant(CloudantTest.java:44)
        at CloudantTest.main(CloudantTest.java:11)
Caused by: java.io.IOException: Failed to read server response from  https://878ff489-3ab9-4d2d-86db-016f19864439-bluemix.cloudantnosqldb.appdomain.cloud:443/_session
        at com.cloudant.http.internal.interceptors.CookieInterceptorBase.wrapIOException(CookieInterceptorBase.java:263)
        at com.cloudant.http.internal.interceptors.CookieInterceptorBase.makeSessionRequest(CookieInterceptorBase.java:164)
        at com.cloudant.http.internal.interceptors.CookieInterceptorBase.requestCookie(CookieInterceptorBase.java:100)
        at com.cloudant.http.internal.interceptors.CookieInterceptorBase.interceptRequest(CookieInterceptorBase.java:176)
        at com.cloudant.http.HttpConnection.execute(HttpConnection.java:291)
        at com.cloudant.client.org.lightcouch.CouchDbClient.execute(CouchDbClient.java:552)
        ... 6 more
Caused by: java.net.ConnectException: Connection timed out (Connection timed out)
        at java.base/java.net.PlainSocketImpl.socketConnect(Native Method)
        at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:399)
        at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:242)
        at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:224)
        at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
        at java.base/java.net.Socket.connect(Socket.java:609)
        at java.base/sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:289)
        at java.base/sun.net.NetworkClient.doConnect(NetworkClient.java:177)
        at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:474)
        at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:569)
        at java.base/sun.net.www.protocol.https.HttpsClient.<init>(HttpsClient.java:265)
        at java.base/sun.net.www.protocol.https.HttpsClient.New(HttpsClient.java:372)
        at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(AbstractDelegateHttpsURLConnection.java:203)
        at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1187)
        at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1081)
        at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:189)
        at java.base/sun.net.www.protocol.http.HttpURLConnection.getOutputStream0(HttpURLConnection.java:1367)
        at java.base/sun.net.www.protocol.http.HttpURLConnection.getOutputStream(HttpURLConnection.java:1342)
        at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getOutputStream(HttpsURLConnectionImpl.java:246)
        at com.cloudant.http.HttpConnection.execute(HttpConnection.java:317)
        at com.cloudant.http.internal.interceptors.CookieInterceptorBase.makeSessionRequest(CookieInterceptorBase.java:140)
        ... 10 more
Error retrieving server response at https://878ff489-3ab9-4d2d-86db-016f19864439-bluemix.cloudantnosqldb.appdomain.cloud:443/account/014aa25148df4eea85b15b3e81ce52d7.
sh-4.4$
```

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,3 +1,6 @@
+# UNRELEASED
+- [FIXED] Enable proxy support for session cookie and IAM authentication.
+
 # 2.20.0 (2021-08-26)
 - [FIXED] Type of `sinceSeq` can be also a `String` besides an `Integer`.
 - [DEPRECATED] This library is now deprecated and will be EOL on Dec 31 2021.

--- a/cloudant-client/src/main/java/com/cloudant/client/api/ClientBuilder.java
+++ b/cloudant-client/src/main/java/com/cloudant/client/api/ClientBuilder.java
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2015, 2020 IBM Corp. All rights reserved.
+ * Copyright © 2015, 2021 IBM Corp. All rights reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
  * except in compliance with the License. You may obtain a copy of the License at
@@ -257,7 +257,7 @@ public class ClientBuilder {
 
             // Create IAM cookie interceptor and set in HttpConnection interceptors
             IamCookieInterceptor cookieInterceptor = new IamCookieInterceptor(this.iamApiKey,
-                    this.url.toString());
+                    this.url.toString(), this.proxyURL);
 
             props.addRequestInterceptors(cookieInterceptor);
             props.addResponseInterceptors(cookieInterceptor);
@@ -274,7 +274,9 @@ public class ClientBuilder {
             //make interceptor if both username and password are not null
 
             //Create cookie interceptor and set in HttpConnection interceptors
-            CookieInterceptor cookieInterceptor = new CookieInterceptor(username, password, this.url.toString());
+            CookieInterceptor cookieInterceptor = new CookieInterceptor(username, password,
+                    this.url.toString(), this.proxyURL);
+
 
             props.addRequestInterceptors(cookieInterceptor);
             props.addResponseInterceptors(cookieInterceptor);

--- a/cloudant-http/src/main/java/com/cloudant/http/internal/interceptors/CookieInterceptor.java
+++ b/cloudant-http/src/main/java/com/cloudant/http/internal/interceptors/CookieInterceptor.java
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2015, 2019 IBM Corp. All rights reserved.
+ * Copyright © 2015, 2021 IBM Corp. All rights reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
  * except in compliance with the License. You may obtain a copy of the License at
@@ -21,6 +21,7 @@ import java.io.IOException;
 import java.io.InputStream;
 import java.io.UnsupportedEncodingException;
 import java.net.HttpURLConnection;
+import java.net.URL;
 import java.net.URLEncoder;
 import java.util.Locale;
 import java.util.logging.Level;
@@ -41,13 +42,26 @@ public class CookieInterceptor extends CookieInterceptorBase {
      * @param baseURL  The base URL to use when constructing an `_session` request.
      */
     public CookieInterceptor(String username, String password, String baseURL) {
+        this(username, password, baseURL, null);
+    }
+
+    /**
+     * Constructs a cookie interceptor with proxy url.
+     * Credentials should be supplied not URL encoded, this class
+     * will perform the necessary URL encoding.
+     *
+     * @param username The username to use when getting the cookie (not URL encoded)
+     * @param password The password to use when getting the cookie (not URL encoded)
+     * @param baseURL  The base URL to use when constructing an `_session` request.
+     * @param proxyURL The URL of the proxy server
+     */
+    public CookieInterceptor(String username, String password, String baseURL, URL proxyURL) {
         // Use form encoding for the user/pass submission
-        super(baseURL, "/_session", "application/x-www-form-urlencoded");
+        super(baseURL, "/_session", "application/x-www-form-urlencoded", proxyURL);
         try {
             this.auth = String.format("name=%s&password=%s", URLEncoder.encode(username, "UTF-8")
-                    , URLEncoder.encode(password, "UTF-8"))
+                            , URLEncoder.encode(password, "UTF-8"))
                     .getBytes("UTF-8");
-            ;
         } catch (UnsupportedEncodingException e) {
             //all JVMs should support UTF-8, so this should not happen
             throw new RuntimeException(e);

--- a/cloudant-http/src/main/java/com/cloudant/http/internal/interceptors/CookieInterceptorBase.java
+++ b/cloudant-http/src/main/java/com/cloudant/http/internal/interceptors/CookieInterceptorBase.java
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2017, 2019 IBM Corp. All rights reserved.
+ * Copyright © 2017, 2021 IBM Corp. All rights reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
  * except in compliance with the License. You may obtain a copy of the License at
@@ -59,13 +59,15 @@ public abstract class CookieInterceptorBase implements HttpConnectionRequestInte
     private final CookieManager cookieManager = new CookieManager();
     private final ReadWriteLock sessionLock = new ReentrantReadWriteLock(true);
     private volatile UUID sessionId = UUID.randomUUID();
+    private final URL proxyURL;
 
     /**
      * @param baseUrl         the server URL to get cookies from
      * @param endpoint        the server endpoint to get cookies from
      * @param requestMimeType the MIME Content-Type to use for the session request
+     * @param proxyURL        the proxy URL
      */
-    protected CookieInterceptorBase(String baseUrl, String endpoint, String requestMimeType) {
+    protected CookieInterceptorBase(String baseUrl, String endpoint, String requestMimeType, URL proxyURL) {
         try {
             baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
             endpoint = endpoint.startsWith("/") ? endpoint : "/" + endpoint;
@@ -77,6 +79,16 @@ public abstract class CookieInterceptorBase implements HttpConnectionRequestInte
             logger.log(Level.SEVERE, "Failed to create URL for session endpoint", e);
             throw new RuntimeException(e);
         }
+        this.proxyURL = proxyURL;
+    }
+
+    /**
+     * @param baseUrl         the server URL to get cookies from
+     * @param endpoint        the server endpoint to get cookies from
+     * @param requestMimeType the MIME Content-Type to use for the session request
+     */
+    protected CookieInterceptorBase(String baseUrl, String endpoint, String requestMimeType) {
+        this(baseUrl, endpoint, requestMimeType, null);
     }
 
     /**
@@ -128,6 +140,10 @@ public abstract class CookieInterceptorBase implements HttpConnectionRequestInte
         HttpConnection conn = Http.POST(url, contentMimeType);
         conn.requestProperties.put("accept", "application/json");
         conn.setRequestBody(payload);
+
+        if (proxyURL != null) {
+            conn.connectionFactory.setProxy(this.proxyURL);
+        }
 
         //when we request the session we need all interceptors except this one
 

--- a/cloudant-http/src/main/java/com/cloudant/http/internal/interceptors/IamCookieInterceptor.java
+++ b/cloudant-http/src/main/java/com/cloudant/http/internal/interceptors/IamCookieInterceptor.java
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2019 IBM Corp. All rights reserved.
+ * Copyright © 2019, 2021 IBM Corp. All rights reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
  * except in compliance with the License. You may obtain a copy of the License at
@@ -34,7 +34,11 @@ public class IamCookieInterceptor extends CookieInterceptorBase {
     private final byte[] tokenRequestPayload;
 
     public IamCookieInterceptor(String apiKey, String baseUrl) {
-        super(baseUrl, "/_iam_session", null);
+        this(apiKey, baseUrl, null);
+    }
+
+    public IamCookieInterceptor(String apiKey, String baseUrl, URL proxyURL) {
+        super(baseUrl, "/_iam_session", null, proxyURL);
 
         // Read iamServer from system property, or set default
         try {

